### ISSUE TITLE
diagnostic for `must_use`

### ISSUE THREAD

**Issue by kjeremy** [rust-analyzer/rust-analyzer#1320]
What would be the process for adding a diagnostic for `must_use`? I just ran into it and would like to add a diagnostic.

**Comment by matklad** [comment]
* we must add attributes to functions in [RawItems](https://github.com/rust-analyzer/rust-analyzer/blob/9800699bab04e97996f0aebec528714165a2619b/crates/ra_hir/src/nameres/raw.rs#L17-L24). We either can add attributes directly to a def, or we may use a side table. Either version is fine.
* when we [create hir::Function](https://github.com/rust-analyzer/rust-analyzer/blob/9800699bab04e97996f0aebec528714165a2619b/crates/ra_hir/src/nameres/collector.rs#L493), we should record the fact that it is `#[must_use]`, the same should be done for ADTs
* after typechecking, in [hir::validation](https://github.com/rust-analyzer/rust-analyzer/blob/9800699bab04e97996f0aebec528714165a2619b/crates/ra_hir/src/expr/validation.rs) we should check that must_use expressions are, in fact, used, and emit diagnostics otherwise.

**Comment by csmoe** [comment]
@matklad like `deny`, `warn`, etc, `must_use` belongs to [diagnostic attributes](https://doc.rust-lang.org/reference/attributes/diagnostics.html#the-must_use-attribute) in general. Not sure diag-attributes are already or will be supported in analyzer.

**Comment by Veykril** [comment]
We should have all the parts in place for this to be implementable now. The relevant part in today's code base for implementing this is now here https://github.com/rust-lang/rust-analyzer/blob/9c0c13ec8e47cfe487f46136b394c0372e957968/crates/hir-ty/src/diagnostics/expr.rs

**Comment by HKalbasi** [comment]
Yes I mean attributes on expression (maybe call others as item attributes? Or it will make confusion with something else?) although if we map our diagnostics to rustc ones (which I think we should do at some point) it wouldn't be a blocker.

**Comment by ChayimFriedman2** [comment]
We have a diagnostic for this now!
